src/OVAL/oval_session.c

```diff
@@ -0,0 +1,657 @@
+	else
+		session->oval.variables = NULL;	/* reset */
+}
+
+void oval_session_set_syschars(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.syschars);
+
+	if (filename != NULL)
+		session->oval.syschars = oscap_source_new_from_file(filename);
+	else
+		session->oval.syschars = NULL;
+}
+
+void oval_session_set_directives(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.directives);
+
+	if (filename != NULL)
+		session->oval.directives = oscap_source_new_from_file(filename);
+	else
+		session->oval.directives = NULL;
+}
+
+void oval_session_set_validation(struct oval_session *session, bool validate, bool full_validation)
+{
+	__attribute__nonnull__(session);
+
+	session->validation = validate;
+	session->full_validation = full_validation;
+}
+
+void oval_session_set_datastream_id(struct oval_session *session, const char *id)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->datastream_id);
+	session->datastream_id = oscap_strdup(id);
+}
+
+void oval_session_set_component_id(struct oval_session *session, const char *id)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->component_id);
+	session->component_id = oscap_strdup(id);
+}
+
+void oval_session_set_results_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.results);
+	session->export.results = oscap_strdup(filename);
+}
+
+void oval_session_set_syschars_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.syschars);
+	session->export.syschars = oscap_strdup(filename);
+}
+
+void oval_session_set_report_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.report);
+	session->export.report = oscap_strdup(filename);
+}
+
+static bool oval_session_validate(struct oscap_source *source, xml_reporter fn, oscap_document_type_t type)
+{
+	if (oscap_source_get_scap_type(source) == type) {
+		if (oscap_source_validate(source, fn, NULL))
+			return false;
+	}
+	else {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Type mismatch: %s. Expecting %s "
+			"but found %s.\n", oscap_source_readable_origin(source),
+			oscap_document_type_to_string(type),
+			oscap_document_type_to_string(oscap_source_get_scap_type(source)));
+		return false;
+	}
+
+	return true;
+}
+
+static int oval_session_load_definitions(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+	__attribute__nonnull__(session->source);
+
+	struct ds_sds_session *sds_session;
+
+	if (!oval_session_validate(session->source, fn, OSCAP_DOCUMENT_SDS)) {
+		/* the source can still be OVAL Definitions */
+		oscap_clearerr();
+		if (!oval_session_validate(session->source, fn, OSCAP_DOCUMENT_OVAL_DEFINITIONS))
+			return 1;
+	}
```

Comment:
Drop `oval_session_validate`.

Just ensure that ```
    oscap_source_get_scap_type(definitions_source) != OSCAP_DOCUMENT_OVAL_DEFINITIONS &&
    oscap_source_get_scap_type(definitions_source) != OSCAP_DOCUMENT_SDS
``` and then run oscap_source_validate.